Zend/zend_hash.c

```diff
@@ -2954,6 +2958,16 @@ ZEND_API void ZEND_FASTCALL zend_hash_sort_ex(HashTable *ht, sort_func_t sort, b
 			zend_hash_rehash(ht);
 		}
 	}
+
+	if (UNEXPECTED(GC_DELREF(ht) == 0)) {
+		zend_array_destroy(ht);
+	} else {
+		/* The array _should_ be added to the gc buffer. However, this causes an
+		 * issue for non-user hashtables that use zend_hash_destroy(), which does
+		 * not remove the destroyed array from the gc buffer, assuming it was
+		 * never added. */
+		// gc_check_possible_root((zend_refcounted *)ht);
```

Comment:
You may try to move this locking code into ``asort()`` and others.

Reply:
We'll have to (re)move `HT_ASSERT_RC1(ht);` from `zend_hash_sort_ex()`, but I think this should work.

Reply:
moving ``HT_ASSERT_RC1(ht);`` is bad :(

Reply:
@dstogov We could introduce another variant of `zend_hash_sort_ex()` that omits it, which can be used explicitly for cases like this. We could also add `HASH_FLAG_ALLOW_COW_VIOLATION`, but since user-code may run during sorting this is likely worse.

Reply:
I think, ``zend_array_sort()`` with the new protection should work.